### ISSUE TITLE
Proposal for parameter server

### ISSUE THREAD

**Issue by agibsonccc** [deeplearning4j/deeplearning4j#1849]
# Few ideas in here:

Stale gradients: Do async updates with a learning rate relative to the timestamp attached to a gradient.  Also applicable for time series

Use rocks db for local synchronization and fast storage

Use zeromq for cross server communication.

Also applicable:
Model Parallelism with GPUs

Why NOT Spark:

It can't be jvm based. Spark can't handle large matrices very well. We need an actual c++ based parameter server for reduced latency as well as considerations around handling cuda the right way.

Edit: With half precision almost supported, there's likely room for us to do some neat stuff around compression if we use nd4j data buffers.

I will update this as I think of more stuff. 

See:
References

[1] Training with intra-block parallel optimization and blockwise model-update filtering. In 2016

IEEE International Conference on Acoustics, Speech and Signal Processing (ICASSP), pages

5880–5884. IEEE, 2016.

[2] Jeffrey Dean, Greg Corrado, Rajat Monga, Kai Chen, Matthieu Devin, Mark Mao, Andrew Senior,

Paul Tucker, Ke Yang, Quoc V Le, et al. Large scale distributed deep networks. In Advances in

Neural Information Processing Systems, pages 1223–1231, 2012.

[3] Augustus Odena. Faster asynchronous sgd. arXiv preprint arXiv:1601.04033, 2016.

[4] Nikko Strom. Scalable distributed dnn training using commodity gpu cloud computing. In Six- teenth Annual Conference of the International Speech Communication Association, 2015. http:

//nikkostrom.com/publications/interspeech2015/strom_interspeech2015.pdf.

[5] Wei Zhang, Suyog Gupta, Xiangru Lian, and Ji Liu. Staleness-aware async-sgd for distributed

deep learning. CoRR, abs/1511.05950, 2015. http://arxiv.org/abs/1511.05950.

**Comment by raver119** [comment]
Support for parallel conversion fp16 -> fp32/64 and fp32/64->fp16 is available for both CUDA and cpu now. So yes, this might be used in network transfers if needed.

**Comment by agibsonccc** [comment]
Parameter server added.
